components/eventbrite/sources/common.js

```diff
@@ -0,0 +1,36 @@
+const eventbrite = require("../eventbrite.app.js");
+
+module.exports = {
+  props: {
+    eventbrite,
+    db: "$.service.db",
+    organization: { propDefinition: [eventbrite, "organization"] },
+  },
+  methods: {
+    generateMeta() {
+      throw new Error("generateMeta is not implemented");
+    },
+    generateEventMeta({ id, name, created }) {
+      return {
+        id,
+        summary: name.text,
+        ts: Date.parse(created),
+      };
+    },
+    emitEvent(data) {
+      const meta = this.generateMeta(data);
+      this.$emit(data, meta);
+    },
+    async paginate(resourceFn, resource, params = null) {
+      let results = [];
+      let hasMoreItems;
+      do {
+        const result = await resourceFn(params);
+        results = results.concat(result[resource]);
+        hasMoreItems = result.pagination.has_more_items;
+        params.continuation = result.pagination.continuation;
+      } while (hasMoreItems);
+      return results;
+    },
```

Comment:
This seems like a good case for a generator, which would:
- Avoid accumulating all the items in a collection before we start processing them (which could put pressure on the memory)
- If an issue happens midway then we'd lose all progress

```suggestion
    async *resourceStream(resourceFn, resource, params = null) {
      let hasMoreItems;
      do {
        const {
          [resource]: items,
          pagination = {},
        } = await resourceFn(params);        
        for (const item in items) {
          yield item;
        }
        
        hasMoreItems = !!pagination.has_more_items;
        params.continuation = pagination.continuation;
      } while (hasMoreItems);
    },
```

Reply:
If you in fact want to retrieve all the items (e.g. to emit an event containing the whole collection), you can accumulate those results where they are actually needed.

For example, you can have something like this in `new-attendee-registered`:
```js
const base = require("./common-webhook");
module.exports = {
  ...base,
  methods: {
    ...base.methods,
    async getData(order) {
      const { id: orderId } = order;
      const attendeesStream = this.resourceStream(
        this.eventbrite.getOrderAttendees.bind(this),
        "attendees",
        orderId,
      );
      const attendees = [];
      for await (const attendee of attendeesStream) {
        attendees.push(attendee);
      }
      // Do the rest...
    },
  },
};
```

For the rest of the cases:

```js
const base = require("../common");
module.exports = {
  ...base,
  hooks: {
    ...base.hooks,
    async deploy() {
      const params = {
        orgId: this.organization,
        params: {
          status: "ended,completed",
        },
      };
      const eventStream = await this.resourceStream(
        this.eventbrite.listEvents.bind(this),
        "events",
        params,
      );
      for await (const event of eventStream) {
        this.emitEvent(event);
      }
    },
  },
};
```

Reply:
Updated to use generator